fix duplicate primary keys for tables without models

sebastienbarre opened the pull request:
This fixes #10354 for CakePHP 2.9

CakeSchema would create duplicate primary keys for tables that already a primary key set on a non-conventional column (e.g. not `id`).

sebastienbarre committed:
fix duplicate primary keys for tables without models

sebastienbarre commented:
Working on fixing coding standard and Postgres test, if I manage to install Postgres.

sebastienbarre committed:
fix coding standard, skip Postgres

sebastienbarre commented:
@markstory Unfortunately this seems unfixable for Postgres, so I had to skip the test for this datasource. Here is the current code in `CakeSchema`:
```php
	protected function _columns(&$Obj) {
		$db = $Obj->getDataSource();
		$fields = $Obj->schema(true);

		$hasPrimaryAlready = false;
		foreach ($fields as $value) {
			if (isset($value['key']) && $value['key'] === 'primary') {
				$hasPrimaryAlready = true;
			}
		}

		$columns = array();
		foreach ($fields as $name => $value) {
			if ($Obj->primaryKey === $name && !$hasPrimaryAlready && !isset($value['key'])) {
				$value['key'] = 'primary';
			}
```
Sadly, the call to `$Obj->schema(true);` is already borked when Postgres is in play. It returns a schema where:
* `id` is *already* hard-coded to be primary (so our new fix below with `$hasPrimaryAlready` doesn't matter),  
* `version_id` (the real primary) is not even marked as `primary` (as it is for the other backends), so that renders the fix even more moot.

The reason for this is this code in `Postgres.php`:
```php
public function describe($model) {
		$table = $this->fullTableName($model, false, false);
		$fields = parent::describe($table);
		$this->_sequenceMap[$table] = array();
		$cols = null;

		if ($fields === null) {
			$cols = $this->_execute(
				"SELECT DISTINCT table_schema AS schema, column_name AS name, data_type AS type, is_nullable AS null,
					column_default AS default, ordinal_position AS position, character_maximum_length AS char_length,
					character_octet_length AS oct_length FROM information_schema.columns
				WHERE table_name = ? AND table_schema = ?  ORDER BY position",
				array($table, $this->config['schema'])
			);

			// @codingStandardsIgnoreStart
			// Postgres columns don't match the coding standards.
			foreach ($cols as $c) {
[...]
				if ($model instanceof Model) {
					if ($c->name === $model->primaryKey) {
						$fields[$c->name]['key'] = 'primary';
```
There you have it, `if ($c->name === $model->primaryKey)` already forces `id` to be primary (for tables that have no explicitly implemented models). Now I try to see if I could loop over all the columns before that, just as I did in `CakeSchema`, to detect if there is a primary already, but no, that information is definitely not made available by `$cols`. To the extent that `version_id` is not even marked as `primary` here. 

Long story short, I don't think Postgres is supporting tables that have no models and use primary keys other than `id`. To my knowledge. This could probably be fixed with some clever joins in Postgres but I don't know anything about that backend.

sebastienbarre commented:
@markstory is there a resident PostgreSQL expert for CakePHP who would want to fix that backend, or are we cool with this solution for 2.9.7? Works for all other backends. Thanks.

markstory commented on lib/Cake/Test/Case/Model/CakeSchemaTest.php:
```diff
@@ -650,6 +683,34 @@ public function testSchemaReadWithCrossDatabase() {
+
+		$hasTable = isset($read['tables']['non_conventional']);
+		$this->assertTrue($hasTable, 'non_conventional table should appear');
+		$versionIdHasKey = $hasTable && isset($read['tables']['non_conventional']['version_id']['key']);
```
You don't need this && here. If hasTable is not true the test will fail and no further assertions will be run.

sebastienbarre replied:
Sure... as long as nothing happens after the assertions, or your test never fails. As it turns out I based my test on the one above, `testSchemaReadWithCrossDatabase()`. In this test, `$fixture->drop($db);` is placed *after* all the assertions. If any assertion fails, the fixture is not dropped and sticks around, and you won't be able to run the test anymore, you will get a `Error: Fixture creation for "non_conventional" failed "SQLSTATE[42S01]: Base table or view already exists: 1050 Table 'non_conventional' already exists"` each time,  until you manually clean the database. That's why I added this `$hasTable` extra test... until I decided to move `$fixture->drop($db);` above all the assertions anyway -- so you are correct they are not needed anymore and I'll remove them. But the test above definitely has that problem...

markstory replied:
I can take care of fixing the other test if you want.

sebastienbarre replied:
Thanks

markstory commented:
I can take a look at why postgres is adding an id column. It might be a few days before I get to it though.

sebastienbarre committed:
clean up assertions now that we are dropping the fixture early

ravage84 commented on lib/Cake/Test/Case/Model/CakeSchemaTest.php:
```diff
@@ -364,6 +364,39 @@ class SchemaCrossDatabaseFixture extends CakeTestFixture {
 }
 
 /**
+ * NonConventionalPrimaryKeyFixture class
```
Nitpicking: No need for "class"

ravage84 replied:
You _could_ add some short explanation what these fixtures are for. So one does not have to look up the related commit.

ravage84 commented on lib/Cake/Test/Case/Model/CakeSchemaTest.php:
```diff
@@ -364,6 +364,39 @@ class SchemaCrossDatabaseFixture extends CakeTestFixture {
+class NonConventionalPrimaryKeyFixture extends CakeTestFixture {
+
+/**
+ * name property
```
Start with upper case. Same for the next two properties.

sebastienbarre replied:
what does?

sebastienbarre committed:
Merge branch '2.x' of https://github.com/cakephp/cakephp into duplicate_primary

sebastienbarre commented:
Hey guys, just checking if this needs more work for 2.9.7 (due in one week?) or we are all good.

markstory commented:
@sebastienbarre I still need to look into why postgres is always adding an id column.

rellect commented on lib/Cake/Model/CakeSchema.php:
```diff
@@ -613,9 +613,16 @@ protected function _columns(&$Obj) {
+		$hasPrimaryAlready = false;
+		foreach ($fields as $value) {
+			if (isset($value['key']) && $value['key'] === 'primary') {
+				$hasPrimaryAlready = true;
```
I'd suggest breaking the loop once the primary is found..

sebastienbarre replied:
Done

sebastienbarre committed:
Merge branch '2.x' of https://github.com/cakephp/cakephp into duplicate_primary

sebastienbarre committed:
optimization, break early

markstory committed:
Update schema reflection for postgres.

Use the schema reflection query from 3.x to provide access to the serial
column attributes. These attributes give a more reliable way to access
primary key data than comparing the primary key of the model. The old
approach failed when schema was generated and concrete model's were
missing.

Refs #10356

sebastienbarre committed:
Merge pull request #2 from cakephp/issue-10356

Update schema reflection for postgres.
